// node-express-ecommerce/src/models/cartsModel.js
import fs from 'fs';
import path from 'path';

const cartsFilePath = path.join(__dirname, '../data/carts.json');

class CartsModel {
    constructor() {
        this.carts = this.loadCarts();
    }

    loadCarts() {
        if (fs.existsSync(cartsFilePath)) {
            const data = fs.readFileSync(cartsFilePath, 'utf-8');
            return JSON.parse(data);
        }
        return [];
    }

    saveCarts() {
        fs.writeFileSync(cartsFilePath, JSON.stringify(this.carts, null, 2));
    }

    getAllCarts() {
        return this.carts;
    }

    getCartById(id) {
        return this.carts.find(cart => cart.id === id);
    }

    createCart(cart) {
        this.carts.push(cart);
        this.saveCarts();
        return cart;
    }

    addProductToCart(cartId, productId) {
        const cart = this.getCartById(cartId);
        if (cart) {
            cart.products.push(productId);
            this.saveCarts();
            return cart;
        }
        return null;
    }
}

export default CartsModel;
