// node-express-ecommerce/src/models/productsModel.js
import fs from 'fs';
import path from 'path';

const productsFilePath = path.join(__dirname, '../data/products.json');

class ProductsModel {
    constructor() {
        this.products = this.loadProducts();
    }

    loadProducts() {
        if (fs.existsSync(productsFilePath)) {
            const data = fs.readFileSync(productsFilePath, 'utf-8');
            return JSON.parse(data);
        }
        return [];
    }

    saveProducts() {
        fs.writeFileSync(productsFilePath, JSON.stringify(this.products, null, 2));
    }

    getAllProducts() {
        return this.products;
    }

    getProductById(id) {
        return this.products.find(product => product.id === id);
    }

    addProduct(product) {
        this.products.push(product);
        this.saveProducts();
    }

    updateProduct(id, updatedProduct) {
        const index = this.products.findIndex(product => product.id === id);
        if (index !== -1) {
            this.products[index] = { ...this.products[index], ...updatedProduct };
            this.saveProducts();
        }
    }

    deleteProduct(id) {
        this.products = this.products.filter(product => product.id !== id);
        this.saveProducts();
    }
}

export default ProductsModel;
